xds/internal/xdsclient/clientimpl.go

```diff
@@ -100,3 +211,151 @@ func (c *clientImpl) close() {
 	}
 	c.logger.Infof("Shutdown")
 }
+
+func (c *clientImpl) getChannelForADS(serverConfig *bootstrap.ServerConfig, callingAuthority *authority) (*xdsChannel, func(), error) {
+	if c.done.HasFired() {
+		return nil, nil, ErrClientClosed
+	}
+
+	initLocked := func(s *channelState) {
+		if c.logger.V(2) {
+			c.logger.Infof("Adding authority %q to the set of interested authorities for channel [%p]", callingAuthority.name, s.channel)
+		}
+		s.interestedAuthorities[callingAuthority] = true
+	}
+	deInitLocked := func(s *channelState) {
+		if c.logger.V(2) {
+			c.logger.Infof("Removing authority %q from the set of interested authorities for channel [%p]", callingAuthority.name, s.channel)
+		}
+		delete(s.interestedAuthorities, callingAuthority)
+	}
+
+	return c.getOrCreateChannel(serverConfig, initLocked, deInitLocked)
+}
+
+func (c *clientImpl) getChannelForLRS(serverConfig *bootstrap.ServerConfig) (*xdsChannel, func(), error) {
+	if c.done.HasFired() {
+		return nil, nil, ErrClientClosed
+	}
+
+	initLocked := func(s *channelState) { s.lrsRefs++ }
+	deInitLocked := func(s *channelState) { s.lrsRefs-- }
+
+	return c.getOrCreateChannel(serverConfig, initLocked, deInitLocked)
+}
+
+func (c *clientImpl) getOrCreateChannel(serverConfig *bootstrap.ServerConfig, initLocked, deInitLocked func(*channelState)) (*xdsChannel, func(), error) {
```

Comment:
Top level comment please, including what the provided callbacks represent.

Reply:
Done.